File: 백준/Silver/1966. 프린터 큐/프린터 큐.cc
#include <iostream>
#include <queue>

using namespace std;

int main()
{
    int t;
    cin >> t;

    while (t--)
    {
        priority_queue<int> pq;
        queue<pair<int, int>> q;

        int n, m;
        cin >> n >> m;

        for (int i = 0; i < n; i++)
        {
            int importance;
            cin >> importance;
            q.push({ i, importance });
            pq.push(importance);
        }

        int count = 0;
        while (!q.empty())
        {
            int index = q.front().first;
            int importance = q.front().second;
            q.pop();

            if (pq.top() == importance)
            {
                pq.pop();
                count++; 

                if (index == m)
                {
                    cout << count << '\n';
                    break;
                }
            }

            else
            {
                q.push({ index, importance });
            }
        }
    }

    return 0;
}

File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {

        Scanner sc = new Scanner(System.in);

        int t = sc.nextInt();

        while (t-- > 0) {
            Queue<Pair<Integer, Integer>> queue = new LinkedList<>();
            PriorityQueue<Integer> priorityQueue = new PriorityQueue<>((o1, o2) -> o2 - o1);

            int n = sc.nextInt();
            int m = sc.nextInt();

            for (int i = 0; i < n; i++) {
                int importance = sc.nextInt();
                queue.add(new Pair<>(i, importance));
                priorityQueue.add(importance);
            }

            int count = 0;
            while (!queue.isEmpty()) {
                int index = queue.peek().first;
                int importance = queue.peek().second;
                queue.remove();

                if (priorityQueue.peek() == importance) {
                    priorityQueue.remove();
                    count++;

                    if (index == m) {
                        System.out.println(count);
                        break;
                    }
                } else {
                    queue.add(new Pair<>(index, importance));
                }
            }
        }
    }
}

class Pair<T, U> {
    public T first;
    public U second;

    public Pair(T first, U second) {
        this.first = first;
        this.second = second;
    }

    @Override
    public String toString() {
        return "{" + first + ", " + second + "}";
    }
}
